cdc/processor/sinkmanager/table_sink_worker.go

```diff
@@ -108,237 +108,25 @@ func validateAndAdjustBound(changefeedID model.ChangeFeedID,
-			zap.Stringer("span", &task.span),
-			zap.Uint64("memory", usedMem-*availableMem))
-		*availableMem = usedMem
-	}
-
-	// Do emit in such situations:
-	// 1. we use more memory than we required;
-	// 2. all events are received.
-	// 3. the pending batch size exceeds maxUpdateIntervalSize;
-	if exceedAvailableMem || allFinished ||
-		needEmitAndAdvance(w.splitTxn, *committedTxnSize, *pendingTxnSize) {
-		if err := w.doEmitAndAdvance(
-			false,
-			events,
-			task,
-			currTxnCommitTs,
-			lastTxnCommitTs,
-			lastPos,
-			batchID,
-			committedTxnSize,
-			pendingTxnSize,
-		); err != nil {
-			return false, errors.Trace(err)
-		}
-		emitted = true
-	}
-
-	// All finished, no need to acquire memory.
-	if allFinished {
-		return emitted, nil
-	}
-
-	if usedMem >= *availableMem {
-		// We just finished a transaction, and the memory usage is still high.
-		// We need try to acquire memory for the next transaction. It is possible
-		// we can't acquire memory, but we finish the current transaction. So
-		// we can wait for next round.
-		if txnFinished {
-			if w.sinkMemQuota.TryAcquire(requestMemSize) {
-				*availableMem += requestMemSize
-				log.Debug("MemoryQuotaTracing: try acquire memory for table sink task",
-					zap.String("namespace", w.changefeedID.Namespace),
-					zap.String("changefeed", w.changefeedID.ID),
-					zap.Stringer("span", &task.span),
-					zap.Uint64("memory", requestMemSize))
-			}
-		} else {
-			// The transaction is not finished and splitTxn is false, we need to
-			// force acquire memory. Because we can't leave rest data
-			// to the next round.
-			if !w.splitTxn {
-				w.sinkMemQuota.ForceAcquire(requestMemSize)
-				*availableMem += requestMemSize
-				log.Debug("MemoryQuotaTracing: force acquire memory for table sink task",
-					zap.String("namespace", w.changefeedID.Namespace),
-					zap.String("changefeed", w.changefeedID.ID),
-					zap.Stringer("span", &task.span),
-					zap.Uint64("memory", requestMemSize))
-			} else {
-				// NOTE: if splitTxn is true it's not required to force acquire memory.
-				// We can wait for a while because we already flushed some data to
-				// the table sink.
-				if err := w.sinkMemQuota.BlockAcquire(requestMemSize); err != nil {
-					return emitted, errors.Trace(err)
-				}
-				*availableMem += requestMemSize
-				log.Debug("MemoryQuotaTracing: block acquire memory for table sink task",
-					zap.String("namespace", w.changefeedID.Namespace),
-					zap.String("changefeed", w.changefeedID.ID),
-					zap.Stringer("span", &task.span),
-					zap.Uint64("memory", requestMemSize))
-			}
-		}
-	}
-	return emitted, nil
-}
-
 func (w *sinkWorker) handleTask(ctx context.Context, task *sinkTask) (finalErr error) {
-	lowerBound, upperBound := validateAndAdjustBound(w.changefeedID, task)
+	advancer := newTableSinkAdvancer(task, w.splitTxn, w.sinkMemQuota, requestMemSize)
+	// The task is finished and some required memory isn't used.
+	defer advancer.cleanup()
```

Comment:
Considering this situation, with splitTxn = true:
1. handles part of a large transaction, commitTs  = 1 and batchID = 1;
2. the task is returned to sinkProgress heap;
3. the table is rescheduled, and advanced to commitTs = 1 and batchID = 1 again.

I think it's possble. To resolve this problem, we can store Advancers in sinkProgress or tableSinkWrapper directly.

Reply:
Or we can make `batchID` as an global atomic variable. I guess it's a better fix in release-6.5.

Reply:
Good catch! I'll fix this bug later! Thanks!